classes/helper.php

```diff
@@ -131,4 +131,50 @@ public static function enable_notifications() : string {
         });
         </script>" . PHP_EOL;
     }
+
+    /**
+     * Merge duplicate local_csp records that have matching sha1hash fields.
+     *
+     * @param array $records Array of local_csp records with matching sha1hash. Must contain at least two records.
+     * @return \stdClass The final object representing a local_csp record once all records are merged into one.
+     */
+    public static function merge_duplicate_records(array $records): \stdClass {
+        global $DB;
+        // If no records provided, there is no object to return so throw exception.
+        if (empty($records)) {
+            throw new \moodle_exception('norecordsfound', 'local_csp');
+        }
+
+        // If only 1 record provided, just return it.
+        if (count($records) === 1) {
+            return reset($records);
+        }
+
+        // Order records by DESC failcounter, as highest fail counter will be the base record.
+        usort($records, function($a, $b) {
+            return $b->failcounter - $a->failcounter;
+        });
+        $baserecord = array_shift($records);
+
+        // Check that sha1hash is identical for all records, then combine fields.
+        foreach ($records as $record) {
+            if ($baserecord->sha1hash != $record->sha1hash) {
+                throw new \moodle_exception('nonduplicaterecords', 'local_csp');
+            }
+            $baserecord->failcounter += $record->failcounter;
+        }
+        $baserecord->timeupdated = time();
+
+        $idstodelete = array_map(function($record) {
+            return $record->id;
+        }, $records);
+
+        // Remove duplicates, and update base record.
+        $transaction = $DB->start_delegated_transaction();
+        $DB->delete_records_list('local_csp', 'id', $idstodelete);
```

Comment:
This might need an explain plan to confirm, but would it be faster to just delete all records with the sha1hash then just insert a new record. My thinking is there is 1 less check per record, and there are tons of records, so it saves on potentially pushing a massive list through.

Reply:
Maybe. It'll take a bit of effort to test this with scale though. I can just do it as your suggestion to avoid having to set up the data. I can't imagine it being slower.